Grove monitor

NickyBobby opened the pull request.

brianrip committed:
Planning database migrations for students associated with locations and activities

brianrip committed:
Comments and ideation for grove monitor model relationships

NickyBobby committed:
Adds google-api-client gem

NickyBobby committed:
Adds Grove Monitor functionality

- Adds Event model user:references location:references start_time:datetime end_time:datetime
- Adds Scan model event:references location:references timestamp:datetime correct:boolean
- Adds scope to find all absent students
- Adds method to find all lost students
- Integration tests for Grove Monitor
- Model tests for events and scans

NickyBobby commented:
- Adds Event model user:references location:references start_time:datetime end_time:datetime
- Adds Scan model event:references location:references timestamp:datetime correct:boolean
- Adds scope to find all absent students
- Adds method to find all lost students
- Integration tests for Grove Monitor
- Model tests for events and scans

@jedschneider this is the start of the grove monitor. Decided to flush out how the database was going to look before we get into grabbing the events from google calendar. Let me know what you think.

jedschneider commented on spec/integration/admin/grove_monitor_spec.rb:
```diff
@@ -2,7 +2,21 @@
+  let(:event2) { student4.events.first }
+  let(:scan) { event.scans.first }
+  let(:scan2) { event2.scans.last }
+
```
when you have this much setup, it might be telling you that you're testing the intricacies of this at the wrong level. I don't see any reason to change this, but I think this indicates that we should be breaking the problem down into smaller chunks and testing those at the unit level. We're pretty light on that side of things in the PR.

jedschneider commented on app/models/users/student.rb:
```diff
@@ -14,4 +18,17 @@ def self.is_student?(data)
+                    result << student if student.events.last.scans.count == 0
+                  end
+
+    result.flatten
```
this must have been fun to do. 🎱

jedschneider replied:
I certainly worry there is not testing around such a specific query. should I be worried?

NickyBobby replied:
I am testing that the scans that fall in between the start and end time of the event, and also have a true value for correct return those students. And the scans that fall outside of the event window do not return the students. The reason why this method looks so janky is because we are checking students that check in to the wrong location and the students that don't check in at all. There definitely could and will be further testing but I'm just getting started with this. There's also gonna be another method to check that the locations match up. But my thinking was that once we confirm that the locations and times match up then I'll change the correct boolean attribute to true. For the purposes of these tests I just made the attribute true or false manually. Also you're right, these queries were SUPER fun to do! 😁 @jedschneider

jedschneider replied:
i don't see these tests under the `model/student_spec` should I be looking elsewhere? Or do you mean, you are currently testing them and haven't pushed?

NickyBobby replied:
@jedschneider Currently I'm testing it as an integration test in `admin/grove_monitor_spec` when you click on lost or absent, we send a filter parameter which brings us back to the `absent` scope and `lost` method within the student model. You're right that they're could be more unit tests in the `student_spec` I'm gonna be working on this today.

jedschneider commented on app/controllers/admin/grove_monitor_controller.rb:
```diff
@@ -1,5 +1,12 @@
 class Admin::GroveMonitorController < ApplicationController
   def show
-    authorize(:student_management, :show?)
+    @students = current_user.grove.students.send(filter_params)
```
[law of demeter](https://en.wikipedia.org/wiki/Law_of_Demeter) says `foo.bar.baz` is a bad idea since you're reaching across object boundaries. And, you can probably do a more performant query if you have a `Teacher.students_in_grove` scope (or a helper class that contextually does the query) that just does a `Student.where(grove_id: self.grove_id)` not really sure how the send is working here, but i trust it does.

jedschneider commented on db/migrate/20160720172751_add_at_school_to_user.rb:
```diff
@@ -0,0 +1,5 @@
+class AddAtSchoolToUser < ActiveRecord::Migration
+  def change
+    add_column :users, :at_school, :boolean
```
it seems like this might not be awesome at some point down the road, but I think we can worry about it at that point. probably a reasonable solution for now.

NickyBobby committed:
Cleans up lost students queries | Adds Student model tests

NickyBobby committed:
Adds new view for all students in Grove

Fixes merge conlicts with master

NickyBobby committed:
Adds location filtering for grove monitor

NickyBobby commented:
@jedschneider I added that teachers can filter the grove monitor by location. The query may change as we push forward and change the way our database structure is set up but it works for the way we currently have it set up. I'll be working more on the database stuff tomorrow.

NickyBobby committed:
Adds seperate index page for grove monitor

Starts Google Calendar Service

jedschneider commented on app/controllers/admin/grove_monitor_controller.rb:
```diff
@@ -1,5 +1,29 @@
+  # /admin/grove_montior?filter=[lost]
+  # /admin/grove_montior?filter=[lost,absent]&teacher=smith
+  # link_to("Lost", admin_grove_monitor_path(filter: [lost]))
+
```
lets remove these comments.

jedschneider commented on app/models/users/student.rb:
```diff
@@ -14,4 +18,22 @@ def self.is_student?(data)
+  end
+
+  def self.checked_in_ids
+    joins(events: [:scans] ).where(at_school: true)
```
you could probably scope this to the grove and reduce the query size.

erinnachen commented on client_secret.json:
```diff
@@ -0,0 +1 @@
+{"installed":{"client_id":"7983846294-o9gd526qhf0nfae1qu0ad0ilf46i517b.apps.googleusercontent.com","project_id":"handy-flame-137623","auth_uri":"https://accounts.google.com/o/oauth2/auth","token_uri":"https://accounts.google.com/o/oauth2/token","auth_provider_x509_cert_url":"https://www.googleapis.com/oauth2/v1/certs","client_secret":"SooeW0faDjaoBaA3O74oM5p3","redirect_uris":["urn:ietf:wg:oauth: ...
```
where did this come from?

NickyBobby replied:
It came from the sample code I used to test out how to work with Google Calendar API. It can be deleted.

NickyBobby committed:
Adds Google Service to grab daily schedule from Google

fix merge conflict

Removed pry from index action

Changes authorization resource for index action of grove monitor

NickyBobby commented:
- Set up Google Service to grab daily schedule from Google Calendar (this is still WIP)
- Grove monitor scopes to events and scans to check whether students are lost or not
- Absent students can also be queried easily from the user model
- Scope functionality may need to be changed as we change the database structure
- Have unit tests for the current scoping so we'll know exactly what to change

NickyBobby committed:
Removes comments

NickyBobby committed:
Adds Presenter to Grove Monitor

NickyBobby committed:
Fixed presenter view issue

NickyBobby commented:
I added a presenter to the grove monitor index view. The style is not great yet.. ahem @Jbern16... But the functionality is there to display students lost/absent/by location. We will have to redo the queries for some of these groups but we'll have a better picture of what that looks like after tomorrow.

NickyBobby committed:
Fixes tests with ambiguous links error

NickyBobby committed:
Deletes grove monitor and grove playlist manager from top nav bar

erinnachen commented on app/policies/grove_monitor_policy.rb:
```diff
@@ -0,0 +1,9 @@
+class GroveMonitorPolicy < ApplicationPolicy
+  def index?
```
This should inherit from TeacherResourcePolicy, and then you won't need these overrides for index and show.

erinnachen commented on app/views/admin/grove_monitor/index.html.erb:
```diff
@@ -0,0 +1,63 @@
+
+  <h2>Absent</h2>
+
+  <% @presenter.absent_students.each do |student| %>
```
It looks like the student presentation is the same in each of these blocks. It would be easier to use a partial with collection:
http://guides.rubyonrails.org/layouts_and_rendering.html
for example:
`<%= render partial: "student", collection: @presenter.absent_students %>`

erinnachen commented on app/presenters/grove_monitor_presenter.rb:
```diff
@@ -0,0 +1,11 @@
+    @locations = {}
+    locations.each do |location|
+      @locations[location.name] = Student.location(location.name) || Student.new
+    end
```
why are we returning a Student.new object instead of empty array?

NickyBobby committed:
Changes grove monitor authorize policy

erinnachen commented on .gitignore:
```diff
@@ -19,3 +19,4 @@
 /coverage
 check_in_tool.iml
 npm-debug.log
+.DS_Store
```
we should probably add the cassettes folder to this.

erinnachen commented on app/controllers/admin/grove_monitor_controller.rb:
```diff
@@ -1,5 +1,23 @@
+
   def show
-    authorize(:teacher_resource, :show?)
+    if params[:name]
```
this logic feels like it should be pushed into the model.

NickyBobby committed:
Extracts partials out of grove monitor index view

NickyBobby committed:
Adds cassettes folder to the gitignore file

NickyBobby committed:
Changes side bar back to it's original style glory

erinnachen commented on app/presenters/grove_monitor_presenter.rb:
```diff
@@ -0,0 +1,11 @@
+class GroveMonitorPresenter
+  attr_reader :lost_students, :absent_students, :locations
+  def initialize(locations)
+    @absent_students = Student.absent
```
This needs to be scoped to the grove.

NickyBobby committed:
Scopes grove monitor presenter to a grove

erinnachen commented:
Cool. This looks GTM 🎉
